core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+     * @return tuple [nextTag, nextEosTag]
+     */
+    auto
+    getTagPositions() {
+        bool        hasTag                = false;
+        std::size_t nextTag               = std::numeric_limits<std::size_t>::max();
+        std::size_t nextEosTag            = std::numeric_limits<std::size_t>::max();
+        bool        asyncEoS              = false;
+        auto        adjust_for_input_port = [this, &hasTag, &nextTag, &nextEosTag, &asyncEoS]<PortLike Port>(Port &port) {
+            if (port.isConnected()) {
+                if constexpr (std::remove_cvref_t<Port>::kIsSynch) {
+                    // get the tag after the one at position 0 that will be evaluated for this chunk.
+                    // nextTag limits the size of the chunk except if this would violate port constraints
+                    nextTag                               = std::min(nextTag, nSamplesUntilNextTag(port, 1).value_or(std::numeric_limits<std::size_t>::max()));
+                    nextEosTag                            = std::min(nextEosTag, samples_to_eos_tag(port).value_or(std::numeric_limits<std::size_t>::max()));
+                    const gr::ConsumableSpan auto tagData = port.tagReader().get(port.tagReader().available());
+                    hasTag |= !tagData.empty() && tagData[0].index == port.streamReader().position() && !tagData[0].map.empty();
+                } else { // async port
+                    if (samples_to_eos_tag(port).transform([&port](auto n) { return n <= port.min_samples; }).value_or(false)) {
+                        asyncEoS = true;
+                    }
+                }
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, inputPorts<PortType::STREAM>(&self()));
+        return std::tuple{hasTag, nextTag, nextEosTag, asyncEoS};
+    }
+
+    /***
+     * skip leftover stride
+     * @param available number of samples that can be consumed from each sync port
+     * @return inputSamples to skip before the chunk
+     */
+    std::size_t
+    inputSamplesToSkipBeforeNextChunk(std::size_t available) {
+        std::size_t inputSamplesToSkipBefore = 0UZ;
+        if constexpr (StrideControl::kEnabled) {                    // check if stride was removed at compile time
+            if (stride.value != 0 && stride.value != denominator) { // check if stride is disabled at runtime
+                if (strideCounter != 0) {                           // drop samples
+                    if (strideCounter >= static_cast<gr::Size_t>(available)) {
+                        inputSamplesToSkipBefore = available;
+                        strideCounter -= static_cast<gr::Size_t>(available);
+                    } else {
+                        inputSamplesToSkipBefore = strideCounter;
+                        strideCounter            = 0;
+                    }
+                }
+            }
+        }
+        return inputSamplesToSkipBefore;
```

Comment:
looks good. I changed from the lambda to a simple bool, since it has no arguments and it is the first thing to be eavauated in the if so I didn't see the benefit of not directly evaluating the boolean expression. Let me know if I overlooked something there.